client/src/main/scala/client/BackPressuredWebSocketClient.scala

```diff
@@ -1,93 +1,103 @@
+    new Observable[String] {
+      override def unsafeSubscribeFn(subscriber: Subscriber[String]): Cancelable = {
+        import subscriber.scheduler
+
+        var webSocket: WebSocket = null
+        try {
+          Utils.log(s"Connecting to $url")
+          webSocket = new WebSocket(url)
+        } catch {
+          case NonFatal(ex) =>
+            Observable.raiseError(ex)
         }
 
-        webSocket.onerror = (event: ErrorEvent) => {
-          downstream.onError(BackPressuredWebSocketClient.Exception(event.message))
+        def closeConnection(): Unit = {
+          if (webSocket != null && webSocket.readyState <= 1)
+            try webSocket.close() catch {
+              case _: Throwable => ()
+            }
         }
 
-        webSocket.onclose = (event: CloseEvent) => {
-          downstream.onComplete()
+        val downstream = Observer.toReactiveSubscriber(subscriber)
+
+        try {
+          webSocket.onopen = (event: Event) => {
+            downstream.onSubscribe(new Subscription {
+              def cancel(): Unit =
+                closeConnection()
+
+              def request(n: Long): Unit = {
+                webSocket.send(n.toString)
+              }
+            })
+          }
+
+          webSocket.onerror = (event: ErrorEvent) => {
+            downstream.onError(BackPressuredWebSocketClient.Exception(event.message))
+          }
+
+          webSocket.onclose = (event: CloseEvent) => {
+            downstream.onComplete()
+          }
+
+          webSocket.onmessage = (event: MessageEvent) => {
+            downstream.onNext(event.data.asInstanceOf[String])
+          }
+        } catch {
+          case NonFatal(ex) =>
+            downstream.onError(ex)
         }
 
-        webSocket.onmessage = (event: MessageEvent) => {
-          downstream.onNext(event.data.asInstanceOf[String])
-        }
-      }
-      catch {
-        case ex: Throwable =>
-          downstream.onError(ex)
+        Cancelable(closeConnection)
       }
     }
 
-  private def closeConnection(webSocket: WebSocket)(implicit s: Scheduler): Unit = {
-    if (webSocket != null && webSocket.readyState <= 1)
-      try webSocket.close() catch { case _: Throwable => () }
-  }
-
-  def onSubscribe(subscriber: Subscriber[String]): Unit = {
+  override def unsafeSubscribeFn(subscriber: Subscriber[String]): Cancelable = {
     import subscriber.scheduler
 
-    var webSocket: WebSocket = null
-    val channel = try {
-      Utils.log(s"Connecting to $url")
-      webSocket = new WebSocket(url)
-      createChannel(webSocket)
-    }
-    catch {
-      case ex: Throwable =>
-        Observable.error(ex)
-    }
-
-    val source = channel.timeout(5.seconds)
-      .doOnCanceled(closeConnection(webSocket))
-
-    source.onSubscribe(new Observer[String] {
-      def onNext(elem: String): Future[Ack] =
-        subscriber.onNext(elem)
+    val subscription: Cancelable =
+      createChannel().subscribe(new Observer[String] {
```

Comment:
This should be an `unsafeSubscribeFn`. Because (1) we know what we are doing and our stuff is protected against contract violations, (2) the normal `subscribe()` wraps subscribers into a `SafeSubscriber` and that's unnecessary when building a new `Observable` (e.g. a new `unsafeSubscribeFn`).